File: javascript/src/day01.js
const fs = require('fs');

function part1() {
  const data = fs.readFileSync('../input/day01.txt', 'utf-8');
  const result = eval(data.replace('\n', ' + '));
  console.log('The final frequency is', result);
}

function part2() {
  const data = fs.readFileSync('../input/day01.txt', 'utf-8');
  const nums = data.split('\n').map(Number);
  const result = findFirstRepeatedFreq(nums);
  console.log('The first repeated frequency is', result);
}

function findFirstRepeatedFreq(nums) {
  const found = new Set();
  let current = 0;
  for(let i = 0; !found.has(current); i = (i + 1) % nums.length) {
    found.add(current);
    current += nums[i];
  }
  return current;
}

module.exports = function() {
  console.log('Day 01');
  part1();
  part2();
}

File: haskell/src/Day01.hs
module Day01 where

import qualified Data.Set as Set

parseSignedNum :: String -> Int
parseSignedNum s =
  let
    sign = head s
    num = read $ tail s
  in
    if sign == '-' then negate num else num

findDuplicateFreq :: [Int] -> Int
findDuplicateFreq freqs = doIt 0 Set.empty (cycle freqs)
  where
    doIt :: Int -> Set.Set Int -> [Int] -> Int
    doIt current previous (f : fs)
      | Set.member current previous = current
      | otherwise = doIt (current + f) (Set.insert current previous) fs


loadInput :: IO [Int]
loadInput = parseInput <$> readFile "../input/day01.txt"
  where
    parseInput = map parseSignedNum . lines

part1 :: IO ()
part1 = do
  freqs <- loadInput
  let
    result = sum freqs
  putStrLn ("The final frequency is " ++ show result)

part2 :: IO ()
part2 = do
  freqs <- loadInput
  let
    firstDuplicate = findDuplicateFreq freqs
  putStrLn ("The first frequency reached twice is " ++ show firstDuplicate)

solve :: IO ()
solve = do
  putStrLn "Day 01"
  part1
  part2
